spring/spring4/spring4-embedded/src/test/java/org/infinispan/spring/provider/SpringCacheCacheTest.java

```diff
@@ -212,6 +209,51 @@ public void testValueLoaderWithNoPreviousValue() {
       assertEquals("test", valueFromCache.get());
    }
 
+   /*
+    * In this test Thread 1 should exclusively block Cache#get method so that Thread 2 won't be able to
+    * insert "thread2" string into the cache.
+    *
+    * The test check this part of the Spring spec:
+    * Return the value to which this cache maps the specified key, obtaining that value from valueLoader if necessary.
+    * This method provides a simple substitute for the conventional "if cached, return; otherwise create, cache and return" pattern.
+    * @see http://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/cache/Cache.html#get-java.lang.Object-java.util.concurrent.Callable-
+    */
+   @Test(timeOut = 30_000)
+   public void testValueLoaderWithLocking() throws Exception {
+      //given
+      CountDownLatch waitUntilThread1LocksValueGetter = new CountDownLatch(1);
+      CountDownLatch waitUntilBothThreadsFinish = new CountDownLatch(2);
+
+      //when
+      fork(() -> {
+         Thread.currentThread().setName("Thread1");
+         cache.get("test", () -> {
+            waitUntilThread1LocksValueGetter.countDown();
+            TimeUnit.MILLISECONDS.sleep(10);
+            return "thread1";
+         });
+         waitUntilBothThreadsFinish.countDown();
+      });
+
+      fork(() -> {
+         Thread.currentThread().setName("Thread2");
+         try {
+            waitUntilThread1LocksValueGetter.await();
+         } catch (InterruptedException e) {
+            throw new AssertionError(e);
+         }
+         cache.get("test", () -> "thread2");
```

Comment:
The problem is that this supplier blocks until the other one finishes. Let me go through the test in steps (perhaps I did some mistake and I just don't understand what you're trying to point me out):

1. Thread 1 and 2 start is undetermined order. They are both supplied to the common executor via `fork` method.
2. Let's assume Thread 2 goes first (just for the sake of explanation).
3. Thread 2 hits `waitUntilThread1LocksValueGetter.await();` and waits (since the latch count is 1).
4. Thread 1 can go now.
5. In supplier method of a Thread 1 unblocks Thread 2 giving it chance to go (`waitUntilThread1LocksValueGetter.countDown();` and `TimeUnit.MILLISECONDS.sleep(10);`).
6. If there was no proper synchronization, the value set in the cache would be "thread2" at this point (set by Thread 2). 
7. But if everything is fine, the implementation will block Thread 2 from progressing because of exclusive lock (at this point Thread 1 already owns write lock and sleeping (sleeping does not release locks)).
8. Thread 1 wakes up and writes the value in cache.
9. Thread 1 releases the lock
10. Thread 2 progresses (since the test is waiting at `thread1.get()` or `thread2.get`).
11. Thread 2 notices that the value is there so it doesn't write anything to the cache.
12. `Cache.ValueWrapper valueAfterGetterIsDone = cache.get("test");` obtains the value which should be `thread1` (since Thread 1 have done the write).

Note that I tried to write this test in pretty generic way. So that I could refactor it in the future.

Does this make sense? Or I made it even worse? :D